### DISCUSSION TITLE
issue: calling setValue from useEffect does not trigger useWatch

### DISCUSSION THREAD

**Discussion by loun4** [react-hook-form/react-hook-form#7728] [Q&A] [answered]
### Version Number

7.25.0

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-js-forked-lbhoo?file=/src/App.js

### Steps to reproduce

1. Go to Codesandbox
2. See the value of name on initial render (null instead of foo)

### Expected behaviour

Calling setValue from an effect should update useWatch values.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment, marked as answer]
Consider lifting `setValue` out of the child or switch the order of `<ComponentA />` and `<ComponentB />` around, so that the component using `useWatch` is before the `setValue` call. It's the expected behavior. See[ Rules section of RHF useWatch docs section](https://react-hook-form.com/api/usewatch#main):
> useWatch execution order matters, which means if you are update form value before the subscription is in place, then the value update will be ignored.

```js
setValue('test', 'data');
useWatch({ name: 'test' }); // ❌ subscription is happened after value update, no update received

useWatch({ name: 'example' }); // ✅ input value update will be received and trigger re-render
setValue('example', 'data'); 
```

**Reply by hokwanhung** [reply]
Thanks for the answer! This saves much debugging time for me.

**Comment by bluebill1049** [comment]
thanks @Moshyfawn! that's correct. you are truly an expert on this lib.

**Comment by loun4** [comment]
Okay, `useWatch ` et `setValue` should be used together at the same level. 
Thanks.

**Comment by Moshyfawn** [comment]
> Okay, useWatch et setValue should be used together at the same level.

Not necessarily. Just make sure `useWatch` is called earlier than `setValue`

**Comment by loun4** [comment]
The codesandbox is updated to show a different case https://codesandbox.io/s/react-hook-form-js-forked-lbhoo?file=/src/App.js

`ComponentB` is now rendered into `ComponentA`, the subscription in the parent `ComponentA` is happened before the value update in `ComponentB` . 
Any idea why the value is not updating in the parent ?

**Reply by bluebill1049** [reply]
children component will be invoked before parent component. At the end of the day, it's all to do with the function execution order.

**Reply by Moshyfawn** [reply]
Refer to [this StackOverflow answer](https://stackoverflow.com/a/58353012/10414512) for some additional info

**Comment by terry623** [comment]
Is there any eslint rule that could prevent this?
